geo/geo.go

```diff
@@ -0,0 +1,246 @@
+		return true
+	}
+	return false
+}
+
+type IPFilter struct {
+	DBPath     string        `json:"dbpath"`
+	DownloadDB bool          `json:"downloadDB"`
+	Allow      []Location    `json:"allow"`
+	Disallow   []Location    `json:"disallow"`
+	Providers  []GeoProvider `json:"-"`
+	mux        sync.Mutex    `json:"-"`
+}
+
+func (f *IPFilter) Empty() bool {
+	if f == nil {
+		return true
+	}
+	for _, loc := range f.Allow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	for _, loc := range f.Disallow {
+		if !loc.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (f *IPFilter) NeedGeoInfo() bool {
+	if f.Empty() {
+		return false
+	}
+	for _, loc := range f.Allow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	for _, loc := range f.Disallow {
+		if len(loc.CountryCode) > 0 || len(loc.Country) > 0 || len(loc.City) > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (f *IPFilter) AllowIP(ip string) (bool, error) {
+	if f.Empty() {
+		return true, nil
+	}
+
+	var loc *Location
+	if f.NeedGeoInfo() {
+		loc = f.GetLocation(ip)
+	} else {
+		loc = &Location{IP: ip}
+	}
+
+	return f.AllowLocation(loc), nil
+}
+
+func (f *IPFilter) GetLocation(ip string) *Location {
+	for _, p := range f.Providers {
+		if p.Ready() {
+			loc := getLocationFromProvider(ip, p.(GeoProvider))
+			if !loc.Empty() {
+				return &loc
+			}
+		}
+	}
+	return &Location{CountryCode: "UNKNOWN", IP: ip}
+}
+
+func (f *IPFilter) AllowLocation(loc *Location) bool {
+	if loc.Empty() {
+		return true
+	}
+
+	for _, l := range f.Disallow {
+		if l.Match(loc) {
+			return false
+		}
+	}
+
+	empty := true
+	for _, l := range f.Allow {
+		if l.Match(loc) {
+			return true
+		}
+		if !l.Empty() {
+			empty = false
+		}
+	}
+
+	return empty
+}
+
+func getLocationFromProvider(ip string, p GeoProvider) Location {
+	loc, err := p.GetLocation(ip)
+	if err != nil {
+		log.Println(err)
+	}
+	return *loc
+}
+
+func GetJsonFile(url, filename string) ([]byte, error) {
+	if s, err := os.Stat(filename); os.IsNotExist(err) || s.Size() == 0 {
+		f, err := os.Create(filename)
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+
+		client := http.Client{
+			Timeout: 5 * time.Second,
+		}
+		resp, err := client.Get(url)
+		if err != nil {
+			os.Remove(filename)
+			return nil, err
+		}
+		defer resp.Body.Close()
+		b, err := ioutil.ReadAll(resp.Body)
+		if !json.Valid(b) {
+			os.Remove(filename)
+			return nil, errors.New("invalid json")
+		}
+		_, err = f.Write(b)
+		if err != nil {
+			os.Remove(filename)
+			return nil, err
+		}
+	}
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (f *IPFilter) AddProvider() {
+	if f.DownloadDB {
+		aws := NewAWSProvider(f.DBPath)
+		err := aws.MaybeUpdate()
```

Comment:
Currently `AddProvider()` will take very long to return, and tuna cannot start to provide service (blocked) before it returns. I think it's better not to call `MaybeUpdate()` in `AddProvider()`, and move `time.Sleep` in `UpdateDataFile()` to the end of for loop so update will not block `AddProvider`